csharp/src/Google.Protobuf/UnknownFieldSet.cs

```diff
@@ -0,0 +1,371 @@
+            }
+            return result;
+        }
+
+        /// <summary>
+        /// Checks if two unknown field sets are equal.
+        /// </summary>
+        public override bool Equals(object other)
+        {
+            if (ReferenceEquals(this, other))
+            {
+                return true;
+            }
+            UnknownFieldSet otherSet = other as UnknownFieldSet;
+            IDictionary<int, UnknownField> otherFields = otherSet.fields;
+            if (fields.Count  != otherFields.Count)
+            {
+                return false;
+            }
+            foreach (KeyValuePair<int, UnknownField> leftEntry in fields)
+            {
+                UnknownField rightValue;
+                if (!otherFields.TryGetValue(leftEntry.Key, out rightValue))
+                {
+                    return false;
+                }
+                if (!leftEntry.Value.Equals(rightValue))
+                {
+                    return false;
+                }
+            }
+            return true;
+        }
+
+        /// <summary>
+        /// Gets the unknown field set's hash code.
+        /// </summary>
+        public override int GetHashCode()
+        {
+            int ret = 1;
+            foreach (KeyValuePair<int, UnknownField> field in fields)
+            {
+                int hash = field.Key.GetHashCode() ^ field.Value.GetHashCode();
+                ret ^= hash;
+            }
+            return ret;
+        }
+
+        #region IMessage Members
+        /// <summary>
+        /// Gets the descriptor. It is never used. Returns null to implement
+        /// the interface member 'IMessage.Descriptor'.
+        /// </summary>
+        public MessageDescriptor Descriptor
+        {
+            get { return null; }
+        }
+
+        /// <summary>
+        /// Checks if the unknown field set is initialized. It is never used.
+        /// Returns true to implement the interface.
+        /// </summary>
+        public bool IsInitialized
+        {
+            get { return true; }
+        }
+        #endregion
+
+        // Optimization:  We keep around the last field that was
+        // modified so that we can efficiently add to it multiple times in a
+        // row (important when parsing an unknown repeated field).
+        private int lastFieldNumber;
+        private UnknownField lastField;
+
+        private UnknownField GetField(int number)
+        {
+            if (lastField != null)
+            {
+                if (number == lastFieldNumber)
+                {
+                    return lastField;
+                }
+            }
+            if (number == 0)
+            {
+                return null;
+            }
+
+            lastField = new UnknownField();
+            UnknownField existing;
+            if (fields.TryGetValue(number, out existing))
+            {
+                lastField.MergeFrom(existing);
```

Comment:
Why are we merging an old one with a new one here? Again, this feels like it's left over from a builder mentality. I'd expect to return `existing` here, and only create a new instance if it's new.

I'd also rename the method to GetOrAdd, as otherwise it's odd that it mutates the set in some cases.

Reply:
done